Adding More Data From AIS

SPDonaghy opened the pull request:
### Description

- Resolves/Related to #29 from local_pathfinding

I modified HelperAISShip.msg by removing:

- speed
- heading

And adding:

- Course over ground (COG)
- Speed over ground (SOG)
- Rate of turn (ROT)
- Dimensions (length and width)

I also added helper msg files for COG, SOG, Dimensions, and ROT and included them in the cmake file.

Aside from general feedback, I will need to know the units this data will come in.

Question for NET: what will be published if the requested data is not available? (ex. if the sog sensor is not working properly on some ship) 

### Verification
I was able to successfully create a HelperAISShip object, with the new data attributes, for use in test_obstacles.py in the local_pathfinding repository.

### Resources
PR for the Issue #29 in local_pathfinding
https://github.com/UBCSailbot/local_pathfinding/pull/34 

I determined what data is available from AIS from [this webpage](https://unstats.un.org/wiki/display/AIS/Overview+of+AIS+dataset)

Information regarding ROT specifically, is from [here](https://documentation.spire.com/ais-fundamentals/rate-of-turn-rot/)

SPDonaghy committed:
Added new helpers and modified HelperAISShip.msg
to include more data from AIS

SPDonaghy committed:
updated COG variable name to cog

SPDonaghy committed:
added new interfaces to CMakeLists.txt
```

SPDonaghy committed:
added rot to AIS ship info

SPDonaghy committed:
added a space

DFriend01 commented:
I think @hhenry01 and @patrick-5546 are the main stakeholders here

patrick-5546 commented on msg/HelperCOG.msg:
Does it have the same convention as `HelperHeading`? If so, we can use that instead of creating a new msg.

SPDonaghy replied:
Yes, they are both measured clockwise from true north

patrick-5546 commented on msg/HelperSOG.msg:
Does it have the same convention as `HelperSpeed`? If so, we can use that instead of creating a new msg.

SPDonaghy replied:
Yes, they are both in knots. So actually I made an error, the units should be knots not km/h

patrick-5546 replied:
We can get NET to convert to kmph on their side before publishing the information. It just makes everything easier if we stick to one unit

SPDonaghy replied:
Okay, in that case maybe they can also just publish ROT in deg/hr as well?

Also for COG and SOG when you say we can use the heading and speed msg files, should I just rename and modify those helper msg files instead of creating the COG and SOG files?

patrick-5546 replied:
> Okay, in that case maybe they can also just publish ROT in deg/hr as well?

Whatever units are conventional and makes the most sense for our codebase

> Also for COG and SOG when you say we can use the heading and speed msg files, should I just rename and modify those helper msg files instead of creating the COG and SOG files?

What modifications would you have to make? Couldn't you use them directly. You can document what COG and SOG is in `HelperAISShip.msg`

SPDonaghy replied:
> Okay, in that case maybe they can also just publish ROT in deg/hr as well?

deg/hr makes most sense to me, as we are using km/h everywhere else.

>What modifications would you have to make? Couldn't you use them directly. You can document what COG and SOG is in HelperAISShip.msg

So in AISShip.msg we would see something like:

#speed over ground in km/h
HelperSpeed sog 

would that be correct?

patrick-5546 replied:
Yes that looks good to me. Deg/hr might make the numerical values pretty small, but I think it should be ok

patrick-5546 commented on msg/HelperROT.msg:
```diff
@@ -0,0 +1,3 @@
+# Unit: -126 to 126 ROT AIS scale corresponds to -708 to 708 degrees per min
```
To me it is a bit unclear what this is. Would this be correct?

```
# Rate of turn
# Unit: degrees per min, increases CW
# Range: [-708, 708)
# To convert to ROT AIS scale [-126, 126): 71 / 15 * sqrt(rot)
```

patrick-5546 replied:
Also you can't square root a negative number

SPDonaghy replied:
Yes that would be correct. I added a link, in the PR description, to the page where I got the info for ROT.
Yes we would need to handle the incoming ROT AIS data carefully to preserve the negative sign when we convert it from [-126, 126) to [-708, 708) for use in our code. This is mentioned in the page I linked in the description also.

Just to clarify, we wont be taking the square root of anything, we will be receiving a value that we need to divide by (71/15), then square, and then preserve the negative sign if the original value was negative.

I realize having the conversion in reverse in the msg file is sort of confusing, so I can change that. As of right now ROT isn't used for anything in the codebase, so I haven't implemented this conversion function anywhere yet.

patrick-5546 reviewed (commented):
Besides the comments below, can you link the resource you used to verify that the added information is provided by AIS in the PR description?

jahn18 commented:
> I think @hhenry01 and @patrick-5546 are the main stakeholders here

@DFriend01 Website depends on these interfaces as well

SPDonaghy committed:
Implemented changes from PR review
Reusing HelperSpeed and HelperHeading messages
instead they will be used for cog and sog
added comments to describe the data in HelperAISSHIP

SPDonaghy committed:
git

patrick-5546 commented on msg/HelperDimensions.msg:
What do you think about making this `HelperDimension.msg` with only one field, then in `HelperAISShip.msg`:
```
HelperDimension width
HelperDimension length
```

SPDonaghy replied:
Sounds good

patrick-5546 reviewed (commented):
Besides the comment below, please update the README accordingly.

You can also resolve the comments above that don't need additional feedback

SPDonaghy committed:
added more description in HelperAISShip

SPDonaghy committed:
modified cmakelist to reflect refactor of
HelperDimensions to HelperDimension

patrick-5546 commented on msg/HelperAISShip.msg:
don't put convention of the fields of type `Helper*` to avoid duplicate information. Source of truth should be the msg definition file.

SPDonaghy replied:
Is there potential for confusion if we are type casting cog as HelperHeading and sog as HelperSpeed without explaining the difference between heading/cog and speed/sog?

SPDonaghy replied:
Or can I still include something like this in HelperAISShip:


\# The ship's course over ground
HelperHeading   cog
\# The ship's speed over ground
HelperSpeed   sog

patrick-5546 replied:
> Or can I still include something like this in HelperAISShip:

Yes this works

patrick-5546 commented on CMakeLists.txt:
Alphabetize

patrick-5546 commented on README.md:
Alphabetize

SPDonaghy committed:
Removed extra comments from HelperAISShip
alphabetized cmakeLists and readme

patrick-5546 committed:
Merge branch 'main' into user/SPDonaghy/add-more-data-from-AIS

patrick-5546 reviewed (approved):
nice

patrick-5546 commented:
You can merge once your local pathfinding PR is ready to merge, as this change will break CI for everything else

SPDonaghy committed:
Merge branch 'main' into user/SPDonaghy/add-more-data-from-AIS
